### ISSUE TITLE
Assertion failure in BlockFrame>>mapInitialIpFrom:to:

### ISSUE THREAD

**Issue by daniels220** [dolphinsmalltalk/Dolphin#238]
It is possible for the textMap of a method to change size when it is compiled for debugging, resulting in an assertion failure trying to convert a block frame homed in that method to a debug frame (for instance, in order to step out of an inner message send into that block frame).

By far the most common cause seems to be that, when compiling for debugging, the Compiler does not make use of the unique, shared, empty block, but compiles a normal clean block in order to insert a breakpoint in it. In doing so it inserts a new textMap entry for the ip of the breakpoint, so the debug textMap has one extra entry. A couple ideas that come to mind:
* Revert to the old behavior (use the special clean block even when debugging), which prevents stepping into that block
* Make a unique, shared empty debug block, equivalent to the normal one but with the extra `#break` bytecode. Or, since performance is relatively unimportant when compiling for debug, compile empty blocks as part of a separate, non-shared CompiledExpression, rather than embedding their bytecodes in the main method.

The one other cause I can see is methods like `AddPoolVariableChange>>changeObject`:

```
changeObject
	| dictionary |
	dictionary := variable isString 
				ifTrue: [Smalltalk at: variable asSymbol]
				ifFalse: [variable].
	^dictionary
```

The temporary `dictionary` is entirely useless, and the non-debug Compiler recognizes this and eliminates it...but it generates an extraneous return bytecode in the process (one in each half of the branch, rather than just one at the end), so it needs two textMap entries corresponding to the same source (the final `^dictionary`). The debug compiler wants to insert a breakpoint between the assignment and the return, so it must generate an actual store-temporary-variable bytecode, and in the process it generates only one return bytecode and one textMap entry for it, meaning the debug textMap has one *less* entry than the normal one.

**Comment by blairmcg** [comment]
Thanks for raising the issue. You've found the two cases detected by an exhaustive test in the long running tests package (see https://github.com/dolphinsmalltalk/Dolphin/search?utf8=%E2%9C%93&q=knownTextMapIssues).
Your also right that the empty block case is the more commonly encountered. It's also the more difficult to correct. The compiler optimisation which replaces an unconditional jump to a return statement with a copy of the return statement is fairly easily fixed by adding a text map entry for unconditional jumps.

**Comment by blairmcg** [comment]
Hi @daniels220, this issue should be fixed for the case of empty blocks now. The fix is in VM 7.0.28. Contrary to what I thought this was much easier to fix than the remaining issue with methods like ``AddPoolVariableChange>>changeObject:``. As the latter is much less common in practice (and a workaround is to remove the redundant temp, e.g. using the inline temp refactoring), I'm not planning on fixing this any time soon. I have however created VM issue https://github.com/dolphinsmalltalk/DolphinVM/issues/95 to track it. Bearing in mind the problem is in the compiler, rather than in the image, I think this Dolphin issue can be closed.

**Comment by daniels220** [comment]
Thanks, that all seems reasonable. If I get around to it, maybe I'll fix the few methods in the image affected by the other issue...
